samplers/aws/xray/internal/manifest.go

```diff
@@ -0,0 +1,359 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	crypto "crypto/rand"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/jinzhu/copier"
+
+	"go.opentelemetry.io/contrib/samplers/aws/xray/internal/util"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+const defaultInterval = int64(10)
+
+const manifestTTL = 3600
+
+// Manifest represents a full sampling ruleset and provides
+// option for configuring Logger, Clock and xrayClient.
+type Manifest struct {
+	Rules                          []Rule
+	SamplingTargetsPollingInterval time.Duration
+	refreshedAt                    int64
+	xrayClient                     *xrayClient
+	clientID                       *string
+	logger                         logr.Logger
+	Clock                          util.Clock
+	mu                             sync.RWMutex
+}
+
+// NewManifest return manifest object configured with logging, clock and xrayClient.
+func NewManifest(addr string, logger logr.Logger) (*Manifest, error) {
+	// generate client for getSamplingRules and getSamplingTargets API call
+	client, err := newClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate clientID for sampling statistics
+	clientID, err := generateClientID()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Manifest{
+		xrayClient:                     client,
+		Clock:                          &util.DefaultClock{},
+		logger:                         logger,
+		SamplingTargetsPollingInterval: 10 * time.Second,
+		clientID:                       clientID,
+	}, nil
+}
+
+// Expired returns true if the manifest has not been successfully refreshed in
+// manifestTTL seconds.
+func (m *Manifest) Expired() bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.refreshedAt < m.Clock.Now().Unix()-manifestTTL
```

Comment:
Just curious, any particular reason on why shouldn't we use Unix timestamp? Basically the [request and response](https://docs.aws.amazon.com/xray/latest/api/API_GetSamplingTargets.html) of `getSamplingTargets` API consist unix timestamp so to avoid having to do back and forth conversion between Unix timestamp and `time.Time` decided to use `Unix` everywhere.

Reply:
A leap second or similar can pass and cause timestamp differences to not be quite accurate so we should avoid Unix as much as possible. This means that the unix timestamps in the API response get converted to `time` (monotonic time) ASAP and not using unix time anymore, and also `Interval` isn't unix time so wouldn't have that issue either.